File: services/Tims.Platform/src/Tims.Application/AccessReview/AccessReviewService.cs
using Tims.Domain.AccessReview;

namespace Tims.Application.AccessReview;

/// <summary>
/// Orchestration mirroring `access-review.service.ts` exactly: `BuildReportAsync` (fetch → kernel →
/// shape → summarize), `AttestAsync` (org-exists → rebuild report → refuse-if-truncated → insert).
/// Deliberately carries NO audit-logging concerns (matches the TS file split: the ROUTER calls
/// `logSecurityEvent`/`logPlatformExport`, not the service) — Task 7's endpoints call
/// <see cref="Tims.Application.Audit.ISecurityEventWriter"/> directly, after calling this service.
/// </summary>
public sealed class AccessReviewService(IAccessReviewRepository repository)
{
    public const int DefaultOrgCap = 10000;

    private readonly IAccessReviewRepository _repository = repository;

    public async Task<AccessReviewReport> BuildReportAsync(
        Guid organizationId, DateTime now, CancellationToken cancellationToken, int cap = DefaultOrgCap)
    {
        var users = await _repository.FetchUsersForReviewAsync(organizationId, cap, cancellationToken).ConfigureAwait(false);
        var truncated = users.Count > cap;
        var rows = (truncated ? users.Take(cap) : users).Select(u => ToRow(u, now)).ToList();

        return new AccessReviewReport(
            Rows: rows,
            Summary: Summarize(rows),
            CrossOrgRoleCount: rows.Count(r => r.Flags.CrossOrgRole),
            Truncated: truncated);
    }

    public async Task<AccessReviewAttestOutcome> AttestAsync(
        Guid organizationId, Guid reviewerId, string? notes, DateTime now, CancellationToken cancellationToken, int cap = DefaultOrgCap)
    {
        if (!await _repository.OrgExistsAsync(organizationId, cancellationToken).ConfigureAwait(false))
        {
            return new AccessReviewAttestOutcome.OrgNotFound();
        }

        var report = await BuildReportAsync(organizationId, now, cancellationToken, cap).ConfigureAwait(false);
        if (report.Truncated)
        {
            return new AccessReviewAttestOutcome.Truncated(cap);
        }

        var summary = report.Summary;
        var attestation = await _repository.InsertAttestationAsync(
            new AccessReviewAttestationInsert(
                organizationId, reviewerId, summary.UserCount, summary.PrivilegedCount,
                summary.StaleCount, summary.DeprovisionGapCount, summary.ExpiredGapCount, notes),
            cancellationToken).ConfigureAwait(false);

        return new AccessReviewAttestOutcome.Success(attestation, summary);
    }

    public Task<IReadOnlyList<AccessReviewAttestationHistoryItem>> ListAttestationsAsync(
        Guid organizationId, int limit, CancellationToken cancellationToken) =>
        _repository.ListAttestationsAsync(organizationId, limit, cancellationToken);

    private static AccessReviewRow ToRow(AccessReviewUserRecord u, DateTime now)
    {
        // Defensive fallback (u.OrganizationId is nullable on the User schema, but the repository's
        // own `where organizationId = @orgId` guarantees every fetched row has one) — mirrors TS's
        // `u.organizationId ?? ''` defensive coalesce, never expected to actually trigger.
        var organizationId = u.OrganizationId ?? Guid.Empty;

        var (status, flags) = AccessRiskKernel.AssessUserAccess(new UserAccessInput(
            organizationId, u.IsActive, u.DeletedAt, u.LastLoginAt,
            u.Roles.Select(r => new RoleAssignment(r.Slug, r.RoleOrganizationId, r.ExpiresAt)).ToList(),
            u.IsPlatformOwner, now));

        return new AccessReviewRow(
            u.Id, $"{u.FirstName} {u.LastName}".Trim(), u.Email, organizationId, u.OrgName, status.ToWire(),
            u.IsPlatformOwner, u.LastLoginAt,
            u.Roles.Select(r => new RoleGrantView(
                r.Slug, r.Name, r.RoleActive, r.AssignedAt, r.AssignedBy, r.CompanyScope, r.UnitScope,
                r.ExpiresAt, r.Grants)).ToList(),
            flags);
    }

    private static AccessReviewSummary Summarize(IReadOnlyList<AccessReviewRow> rows) => new(
        UserCount: rows.Count,
        PrivilegedCount: rows.Count(r => r.Flags.Privileged),
        StaleCount: rows.Count(r => r.Flags.Stale),
        DeprovisionGapCount: rows.Count(r => r.Flags.DeprovisionGap),
        ExpiredGapCount: rows.Count(r => r.Flags.ExpiredGrant));
}

/// <summary>Outcome of <see cref="AccessReviewService.AttestAsync"/> — the endpoint (Task 7) pattern-matches
/// this to a status code (200/404/412), matching this codebase's established outcome-type convention
/// (e.g. <c>SuccessionWriteUseCase</c>'s null-return pattern) rather than throwing for expected outcomes.</summary>
public abstract record AccessReviewAttestOutcome
{
    public sealed record Success(AccessReviewAttestation Attestation, AccessReviewSummary Summary) : AccessReviewAttestOutcome;

    public sealed record OrgNotFound : AccessReviewAttestOutcome;

    public sealed record Truncated(int Cap) : AccessReviewAttestOutcome;
}

File: packages/api/src/services/access-review.service.ts
import { TRPCError } from '@trpc/server';
import { accessReviewRepository } from '../repositories/access-review.repository';
import { assessUserAccess, type AccessStatus, type AccessRiskFlags } from '../access/access-review-kernel';

// CB-2b — access-review orchestration: apply the risk kernel, shape rows, compute the
// per-org summary that feeds a recertification attestation. Pure of tRPC/audit concerns
// (the router audits). `now` is injected for deterministic tests.

export interface RoleGrantView {
  slug: string;
  name: string;
  roleActive: boolean;
  assignedAt: Date;
  assignedBy: string | null;
  companyScope: string | null;
  unitScope: string | null;
  expiresAt: Date | null;
  grants: string[]; // 'module:action:scope'
}

export interface AccessReviewRow {
  userId: string;
  name: string;
  email: string;
  organizationId: string;
  orgName: string | null;
  status: AccessStatus;
  isPlatformOwner: boolean;
  lastLoginAt: Date | null;
  roles: RoleGrantView[];
  flags: AccessRiskFlags;
}

export interface AccessReviewSummary {
  userCount: number;
  privilegedCount: number;
  staleCount: number;
  deprovisionGapCount: number;
  expiredGapCount: number;
}

export interface AccessReviewReport {
  rows: AccessReviewRow[];
  summary: AccessReviewSummary;
  /** Data-integrity anomaly count (grant corruption) — surfaced, not part of the attestation snapshot. */
  crossOrgRoleCount: number;
  /** true when more users exist than the cap (rows are the first `cap`) — honest, no silent truncation. */
  truncated: boolean;
}

// The review is always ORG-SCOPED (the attestation/export/audit unit). One org is
// bounded per-tenant, so a high safety cap guards memory; `attest` REFUSES a truncated
// org rather than persist under-counted (false) evidence.
const ORG_CAP = 10000;

type ReviewUser = Awaited<ReturnType<typeof accessReviewRepository.fetchUsersForReview>>[number];

function toRow(u: ReviewUser, now: Date): AccessReviewRow {
  const { status, flags } = assessUserAccess({
    organizationId: u.organizationId ?? '',
    isActive: u.isActive,
    deletedAt: u.deletedAt,
    lastLoginAt: u.lastLoginAt,
    roles: u.userRoles.map((ur) => ({
      slug: ur.role.slug,
      organizationId: ur.role.organizationId,
      expiresAt: ur.expiresAt,
    })),
    isPlatformOwner: u.isPlatformOwner,
    now,
  });
  return {
    userId: u.id,
    name: `${u.firstName ?? ''} ${u.lastName ?? ''}`.trim(),
    email: u.email,
    organizationId: u.organizationId ?? '',
    orgName: u.organization?.name ?? null,
    status,
    isPlatformOwner: u.isPlatformOwner,
    lastLoginAt: u.lastLoginAt,
    roles: u.userRoles.map((ur) => ({
      slug: ur.role.slug,
      name: ur.role.name,
      roleActive: ur.role.isActive,
      assignedAt: ur.assignedAt,
      assignedBy: ur.assignedBy,
      companyScope: ur.companyScope,
      unitScope: ur.unitScope,
      expiresAt: ur.expiresAt,
      grants: ur.role.rolePermissions.map((rp) => `${rp.permission.module}:${rp.permission.action}:${rp.scope}`),
    })),
    flags,
  };
}

function summarize(rows: AccessReviewRow[]): AccessReviewSummary {
  return {
    userCount: rows.length,
    privilegedCount: rows.filter((r) => r.flags.privileged).length,
    staleCount: rows.filter((r) => r.flags.stale).length,
    deprovisionGapCount: rows.filter((r) => r.flags.deprovisionGap).length,
    expiredGapCount: rows.filter((r) => r.flags.expiredGrant).length,
  };
}

export const accessReviewService = {
  async buildReport(organizationId: string, now: Date): Promise<AccessReviewReport> {
    const users = await accessReviewRepository.fetchUsersForReview(organizationId, ORG_CAP);
    const truncated = users.length > ORG_CAP;
    const rows = (truncated ? users.slice(0, ORG_CAP) : users).map((u) => toRow(u, now));
    return {
      rows,
      summary: summarize(rows),
      crossOrgRoleCount: rows.filter((r) => r.flags.crossOrgRole).length,
      truncated,
    };
  },

  /**
   * Record a per-org recertification: recompute the org's snapshot and persist it as
   * durable CC6.2–6.3 evidence. REFUSES a truncated org (would persist under-counted =
   * false compliance evidence). Returns the attestation + the summary it was built from.
   */
  async attest(
    organizationId: string,
    reviewerId: string,
    notes: string | null,
    now: Date,
  ): Promise<{
    attestation: Awaited<ReturnType<typeof accessReviewRepository.insertAttestation>>;
    summary: AccessReviewSummary;
  }> {
    if (!(await accessReviewRepository.orgExists(organizationId))) {
      throw new TRPCError({ code: 'NOT_FOUND', message: 'Organizacion no encontrada' });
    }
    const report = await this.buildReport(organizationId, now);
    if (report.truncated) {
      throw new TRPCError({
        code: 'PRECONDITION_FAILED',
        message: `La organizacion excede ${ORG_CAP} usuarios; no se puede certificar automaticamente sin subcontar`,
      });
    }
    const { summary } = report;
    const attestation = await accessReviewRepository.insertAttestation({
      organizationId,
      reviewerId,
      userCount: summary.userCount,
      privilegedCount: summary.privilegedCount,
      staleCount: summary.staleCount,
      deprovisionGapCount: summary.deprovisionGapCount,
      expiredGapCount: summary.expiredGapCount,
      notes,
    });
    return { attestation, summary };
  },
};
